vignettes/performance.Rmd

```diff
@@ -0,0 +1,223 @@
+---
+title: "Optimize `polars` performance"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Optimize `polars` performance}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+options(rmarkdown.html_vignette.check_title = FALSE)
+```
+
+
+As highlighted by the [DuckDB benchmarks](https://duckdblabs.github.io/db-benchmark/),
+`polars` is very efficient to deal with large datasets. Still, one can make `polars` 
+even faster by following some good practices.
+
+
+# Lazy vs eager execution
+
+## Order of operations
+
+In the "Get Started" vignette, we mostly used *eager* execution. This is the
+classic type of execution in R: write some code, link functions with pipes, run
+the code and it is executed line by line.
+
+Most of the time, this kind of execution is perfectly fine. With datasets of a
+reasonable size (let's say up to a few hundreds of thousands of observations and
+a few dozens columns), we don't really have to worry about whether our code is
+optimized to save memory and time.
+
+However, when we start dealing with much larger datasets, the order in which
+functions are applied is extremely important. Indeed, some functions are much
+more memory intensive than others.
+
+For instance, let's say we have a dataset with 1M observations containing
+country names and a few numeric columns. We would like to keep just a few of
+these countries and sort them alphabetically. Here, we have two operations:
+filtering and sorting. Sorting is much harder to do than filtering. To filter
+data, we simply check whether each row fills some conditions, but to sort data
+we have to compare rows between them and rearrange them as we go. If we don't
+take this into account, sorting data before filtering it can slow down our
+pipeline significantly.
+
+```{r}
+countries = c("France", "Germany", "United Kingdom", "Japan", "Columbia", 
+               "South Korea", "Vietnam", "South Africa", "Senegal", "Iran")
+
+set.seed(123)
+test = data.frame(
+  country = sample(countries, 1e6, TRUE),
+  x = sample(1:100, 1e6, TRUE),
+  y = sample(1:1000, 1e6, TRUE)
+)
+
+bench::mark(
+  sort_filter = {
+    tmp = test[order(test$country), ]
+    subset(tmp, country %in% c("United Kingdom", "Japan", "Vietnam"))
+  },
+  filter_sort = {
+    tmp = subset(test, country %in% c("United Kingdom", "Japan", "Vietnam"))
+    tmp[order(tmp$country), ]
+  }
+)
+```
```

Comment:
I don't know if it is a good idea to do the benchmarking on a GitHub runner every time it is deployed.

Reply:
True, maybe I can cache the results of the benchmarks that are not really related to `polars` (like this one) but those that compare eager and lazy queries we want their timings to be up to date, right?

Reply:
In the end I just put all current results and do not evaluate the benchmarks. They should probably be updated from time to time